cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/DefaultTaskSystemAppContext.java

```diff
@@ -0,0 +1,219 @@
+ */
+public class DefaultTaskSystemAppContext implements TaskSystemAppContext, ServiceDiscoverer,
+  SecureStore {
+
+  private final Admin admin;
+  private final PreferencesFetcher preferencesFetcher;
+  private final CConfiguration cConf;
+  private final ClassLoader artifactClassLoader;
+  private final PluginFinder pluginFinder;
+  private final ArtifactId artifactId;
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final SecureStore secureStore;
+  private final String artifactNameSpace;
+  private final ArtifactManager artifactManager;
+  private final RetryStrategy retryStrategy;
+  private final String serviceName;
+
+  DefaultTaskSystemAppContext(CConfiguration cConf, DatasetFramework dsFramework,
+                              SecureStoreManager secureStoreManager, MessagingService messagingService,
+                              RetryStrategy retryStrategy, NamespaceQueryAdmin namespaceQueryAdmin,
+                              PreferencesFetcher preferencesFetcher, PluginFinder pluginFinder,
+                              DiscoveryServiceClient discoveryServiceClient, SecureStore secureStore,
+                              ArtifactManagerFactory artifactManagerFactory, String serviceName,
+                              String artifactNameSpace, ArtifactId artifactId, ClassLoader artifactClassLoader) {
+    this.cConf = cConf;
+    this.artifactNameSpace = artifactNameSpace;
+    NamespaceId namespaceId = new NamespaceId(artifactNameSpace);
+    this.admin = new DefaultAdmin(dsFramework, namespaceId, secureStoreManager,
+                                  new BasicMessagingAdmin(messagingService, namespaceId),
+                                  retryStrategy, null, namespaceQueryAdmin);
+    this.retryStrategy = retryStrategy;
+    this.preferencesFetcher = preferencesFetcher;
+    this.artifactClassLoader = artifactClassLoader;
+    this.pluginFinder = pluginFinder;
+    this.artifactId = artifactId;
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.artifactManager = artifactManagerFactory.create(namespaceId, retryStrategy);
+    this.secureStore = secureStore;
+    this.serviceName = serviceName;
+  }
+
+  @Override
+  public Admin getAdmin() {
+    return admin;
+  }
+
+  @Override
+  public Map<String, String> getPreferencesForNamespace(String namespace, boolean resolved) throws IOException {
+    try {
+      return preferencesFetcher.get(new NamespaceId(namespace), resolved).getProperties();
+    } catch (NotFoundException nfe) {
+      throw new IllegalArgumentException(String.format("Namespace '%s' does not exist", namespace), nfe);
+    }
+  }
+
+  @Override
+  public PluginConfigurer createPluginConfigurer(String namespace) throws IOException {
+    File tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    //Create the same directory every time for caching
```

Comment:
The part that is not fully clear to me is - if it has plugins from multiple restarts, does it mean, at some point it could have some malicious plugins and will not serve the purpose of pod isolation?
since plugin instantiator can instantiate malicious plugins? Or thats not a possibility.

Reply:
Changed the implementation to use a temp directory and clean up after task. PluginInstantiator uses this directory to symlink from the original location and unpack jar file to load classes. This can be optimized to use the unpacked folder directly once the end point is available. @wyzhang